vircadia/vircadia
# Entity update improvements 2

also fixes https://github.com/kasenvr/project-athena/issues/681

The current rendering code is a mess of locks and threads and chaos.  This is an attempt to disentangle some things, hopefully resulting in some performance improvements.

The general idea is:
- EntityItems must be threadsafe as they are accessed by multiple threads (main thread, script threads, render thread)
- RenderableEntityItems operate on two threads, the main thread (doRenderUpdateSynchronous, needsRenderUpdate) and the render thread (doRenderUpdateAsynchronous, doRender)
- RenderableEntityItems should do as much work as possible in doRenderUpdateAsynchronous so that we don't have to lock as much.
- if a property is accessed on both the main thread and the render threads, it needs to be locked properly
- needsRenderUpdate is ONLY called when an EntityItem is edited.  if a Renderable needs to update constantly to check for something (like textures loading or model animation), it just needs to `emit requestRenderUpdate()` and needsRenderUpdate can be removed

Test plan:
- To test the model dimensions/collision bug fix, run this script: https://gist.githubusercontent.com/HifiExperiments/5b3a0207e222546b27ed182244a6861a/raw/cce164f967f7632cdccc845b75028e64ee1ef4ce/ModelTest.js
- Press `o` to toggle the modelURL.  Verify that the dimensions and the collision shape are correct.
- Confirm the test plan from #850 works
- For the rest of the PR, confirm that modifying various properties that effect rendering still work.  In particular, for every entity type (but especially models), confirm that the following still work (as they do in master, some entities don't support these properties): visibility, shadows, render layer, primitive mode, renderWithZones, position/orientation/scale
- confirm that models parented to your head joint properly cauterize (don't render) when you're in first person, but do render normally otherwise
